lib/AST/TypeSubstitution.cpp

```diff
@@ -220,14 +220,27 @@ operator()(CanType dependentType, Type conformingReplacementType,
 ProtocolConformanceRef LookUpConformanceInSubstitutionMap::
 operator()(CanType dependentType, Type conformingReplacementType,
            ProtocolDecl *conformedProtocol) const {
+  auto result = Subs.lookupConformance(dependentType, conformedProtocol);
+  if (!result.isInvalid())
+    return result;
+
   // Lookup conformances for archetypes that conform concretely
   // via a superclass.
   if (auto archetypeType = conformingReplacementType->getAs<ArchetypeType>()) {
     return conformedProtocol->getModuleContext()->lookupConformance(
         conformingReplacementType, conformedProtocol,
         /*allowMissing=*/true);
   }
-  return Subs.lookupConformance(dependentType, conformedProtocol);
+
+  // Otherwise, the original type might be fixed to a concrete type in
+  // the substitution map's input generic signature.
+  if (Subs.getGenericSignature()->isConcreteType(dependentType)) {
+    return conformedProtocol->getModuleContext()->lookupConformance(
+        conformingReplacementType, conformedProtocol,
+        /*allowMissing=*/true);
+  }
+
+  return ProtocolConformanceRef::forInvalid();
```

Comment:
Huh, so if the `dependentType` is not concrete, we don't need to check the substitution map for a conformance like we did before? It can't be generic but carry a conformance constraint in the sub map's generic signature?

Otherwise, looks like we could fold these two cases into just `conformingReplacementType->is<ArchetypeType>() || ... is concrete ...`. I also see the value in having them separate to explain both cases.

Reply:
> Huh, so if the dependentType is not concrete, we don't need to check the substitution map for a conformance like we did before? It can't be generic but carry a conformance constraint in the sub map's generic signature?

The lookup into the substitution map is still there, we just do it _first_ and then go the fallback logic if the lookup failed. lookupConformance() is completely robust against invalid inputs, but in the happy path it can avoid lazily building the requirement machine in easy cases. This way, we avoid the isConcreteType() check in the happy case, which would have to build the requirement machine.